library/internal/buffer.go

```diff
@@ -0,0 +1,64 @@
+package internal
+
+import (
+	"bytes"
+	"io"
+)
+
+//MultiplexingBuffer is a type that will allow storing, replaying and streaming data from it's internal buffer
+//It implements io.Writer storing the written bytes to an unbound buffer internally
+//The Reader function will return an io.Reader that will stream data starting with the buffer, until the Close method of the underlying MultiplexingBuffer is called
+type MultiplexingBuffer struct {
+	closed bool
+	buffer []byte
```

Comment:
race condition.  For example when Read/Write/Close is called by two different goroutines.